server/utils/email/digest.ts

```diff
@@ -0,0 +1,81 @@
+import pick from 'lodash.pick';
+import omit from 'lodash.omit';
+import { GroupedActivityItems } from 'client/components/Email';
+
+import { ActivityItem } from 'types';
+import { fetchActivityItems } from 'server/activityItem/fetch';
+
+const getAffectedObject = (item) =>
+	item.pubId
+		? { id: item.pubId, title: item.payload.pub.title }
+		: item.collectionId
+		? { id: item.collectionId, title: item.payload.collection.title }
+		: 'page' in item.payload
+		? pick(item.payload.page, ['id', 'title'])
+		: { id: item.communityId, title: item.payload.community.title };
+
+const getAffectedObjectIcon = (item: ActivityItem) =>
+	item.pubId
+		? 'pubDoc'
+		: item.collectionId
+		? 'collection'
+		: 'page' in item.payload && item.payload.page.id
+		? 'page-layout'
+		: 'office';
+
+type KeyedActivityItem = ActivityItem & {
+	displayKey: string;
+};
+
+export const getDigestData = async (initialData) => {
+	const {
+		communityData: { id: communityId },
+		scopeData: { scope },
+		loginData: { id: userId },
+	} = initialData;
+	const since = new Date(Date.now() - 7 * 24 * 60 * 60 * 1000).toISOString();
+
+	const { activityItems, associations } = await fetchActivityItems({ scope, limit: 150, since });
+	const activityItemsGroupedByObjectId: Record<
+		string,
+		KeyedActivityItem[]
+	> = activityItems.reduce((result, item) => {
+		const objectId = getAffectedObject(item).id;
+		const payloadKeys = Object.keys(item.payload)
+			.sort()
+			.join();
+		const displayKey = `${item.kind} - ${payloadKeys}`;
+		return {
+			...result,
+			[objectId]: [...(result[objectId] || []), { ...item, displayKey }],
+		};
+	}, {});
+	const dedupedActivityItems = Object.keys(activityItemsGroupedByObjectId).reduce(
+		(memo, objectId) => ({
+			...memo,
+			[objectId]: {
+				items: activityItemsGroupedByObjectId[objectId]
+					.sort((first, second) => (first.timestamp > second.timestamp ? -1 : 1))
+					.reduce(
+						(result, item) =>
+							item.displayKey in result
+								? result
+								: { ...result, [item.displayKey]: item },
+						{},
+					),
+				title: getAffectedObject(activityItemsGroupedByObjectId[objectId][0])?.title,
+				icon: getAffectedObjectIcon(activityItemsGroupedByObjectId[objectId][0]),
+			},
+		}),
+		{},
+	);
+
+	const communityItems: GroupedActivityItems = pick(dedupedActivityItems, communityId);
+	const pubItems: GroupedActivityItems = omit(dedupedActivityItems, communityId);
```

Comment:
This is the meat and potatoes of the affair, sorting and grouping activity items so that they can be displayed as: "Here's all the distinct kinds of things that have happened to pub *ABCD* in the last week!"